lib/src/main/java/com/cloudbees/groovy/cps/impl/BlockScopeEnv.java

```diff
@@ -12,16 +13,34 @@
  */
 // TODO: should be package local once all the impls move into this class
 public class BlockScopeEnv extends ProxyEnv {
-    private final Map<String,Object> locals = new HashMap<String, Object>();
-    private Map<String, Class> types = new HashMap<String, Class>();
+    private Map<String,Object> locals;
+    private Map<String, Class> types;
 
     public BlockScopeEnv(Env parent) {
+        this(parent, 0);
+    }
+
+    public BlockScopeEnv(Env parent, int localsSize) {
         super(parent);
+        if (localsSize <= 0) {
+            locals = Collections.emptyMap();
+            types = Collections.emptyMap();
+        } else {
+            locals = new HashMap<String, Object>(localsSize);
+            types = new HashMap<String, Class>(localsSize);
+        }
     }
 
     public void declareVariable(Class type, String name) {
+        if (locals == Collections.EMPTY_MAP) {
```

Comment:
Awkward, and fragile since you could accidentally try to write to it, resulting in an `UnsupportedOperationException` which would be confusing.

Why not just leave it null unless and until a request is made to store something in it? You need then only do a null check in a few places, deleting `getTypes`.

Reply:
It's a private field, exposed only by this trivial set of setters, how on early would I "accidentally" try to write to it?

Reply:
Just a stylistic suggestion to make the code more legible (IMO) to the next person to edit this code, perhaps a future version of yourself. At least a comment explaining that this non-null collection field is sometimes mutable and sometimes not would be in order.

The typical idiom is something like

```java
/** Null unless and until something needs to be stored in it. */
@CheckForNull private Map<String,Object> stuff;
@CheckForNull public Object getStuff(String key) {
    return stuff != null ? stuff.get(key) : null;
}
public void putStuff(String key, Object value) {
    if (stuff == null) {
        stuff = new HashMap<>();
    }
    stuff.put(key, value);
}
```

and as noted you can then get rid of `getTypes` and special deserialization logic, because there is no difference between a newly constructed object with a null field, and a deserialized old object with a null field.

BTW if there are some subclasses for which you can know for sure that the collection is always empty, or always contains a single element, then you could optimize further by removing the field from an `abstract` superclass and making the getter-like method `abstract`. One implementation could keep a lazily-initialized `HashMap` field to support any number of entries; another could keep single `final` fields for the key and value. Works best if the setter-like methods can be pushed down into certain concrete subclasses, perhaps narrowing the declared types used by some callers to reflect statically known information about the particular implementation in use. Not sure offhand if that applies here—would need to study the code.